Boost: Add first user flow

haqadn opened the pull request:
Fixes 142-gh-Automattic/boost-cloud

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
- [x] Added a pricing table page that compares between the free and paid plan of boost
- [x] User can continue to the purchase flow by clicking the purchase button on the new page
- [x] The purchase page will automatically appear for first-time users
- [x] Once a user chooses a plan, the purchase plan will no longer appear
- [x] Rewrite the pricing component in svelte
- [x] Plug in pricing from real data source

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion
p1HpG7-hBQ-p2 

#### Does this pull request change what data or activity we track or use?
No
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

#### Testing instructions:
* For now, you need to manually access the page from: `wp-admin/admin.php?page=jetpack-boost#/getting-started`
* Set option `jb_get_started` to true to force the user to redirect to the get started page every time they access boost dashboard.
* Choose a plan, and the forced redirect to the getting started page should go away.

##### Notes
- The CTA on free plan will permanently disable the getting started page.
- Choosing the paid plan will not disable the getting started page, it will only redirect to the purchase flow.
- Accessing the purchase success page(`wp-admin/admin.php?page=jetpack-boost#/purchase-successful` will disable the getting started page.

haqadn committed:
Add primary implementation for pricing table

haqadn committed:
Refactor pricing table style

haqadn committed:
Improve comparison table style

haqadn committed:
Improve responsive style

haqadn committed:
Fix scroll-bar issue

haqadn committed:
Fix pop-out issue

haqadn committed:
Stich together the getting started flow

haqadn committed:
changelog

haqadn committed:
Fix on connection complete

Previously it was not being called.

haqadn committed:
Force redirect to getting started on connection

haqadn committed:
Change changelog type to patch

haqadn committed:
Get rid of unused variable warning in Get_Started enadpoint

haqadn committed:
Fix uninstall

haqadn committed:
Fix critical-css-advanced path

haqadn committed:
Fix unit tests

haqadn committed:
Rewrite ProductPrice component in svelte

haqadn committed:
Plug in real data

thingalon commented on projects/plugins/boost/app/assets/src/css/components/comparison-table.scss:
```diff
@@ -0,0 +1,160 @@
+            border-top: 0;
+            border-bottom: 1px solid $gray_5;
+        }
+    }
```
Inconsistent indentation throughout this file; mix of spaces and tabs.

thingalon commented on projects/plugins/boost/app/assets/src/css/components/context-info.scss:
```diff
@@ -0,0 +1,88 @@
+
+	margin-left: calc(12 / 18 * var( --icon-size ));
+
+	&__icon{
```
mixed spaces and tabs here too.

thingalon commented on projects/plugins/boost/app/assets/src/css/components/price.scss:
```diff
@@ -0,0 +1,72 @@
+    font-weight: 700;
+    line-height: 52px;
+    margin: 0;
+    padding: 0;
```
mixed here, too.

thingalon commented on projects/plugins/boost/app/assets/src/css/main/common.scss:
```diff
@@ -27,3 +27,18 @@ $definition-combinations: (
+}
+
+.hidden-md {
+  @include breakpoint( 'md' ) {
```
here too

haqadn replied:
I've fixed indentation in all scss files.

thingalon commented on projects/plugins/boost/app/assets/src/js/Main.svelte:
```diff
@@ -14,14 +15,15 @@
+	<Route path="/*" let:location let:navigate>
 		<div id="jb-settings" class="jb-settings jb-settings--main">
 			<div class="jb-container">
 				<Header />
```
🤔 This structure seems a little strange to me.

Are `upgrade`, `purchase-successful` and `getting-started` available to users before they connect to Jetpack? It seems like the if/else which forces users to the connection screen belongs at a higher level in this structure.

It also seems a bit weird that `<header />` is added to the settings and connections pages in this structure here, but `getting-started` adds a `<header />` itself.

I understand if fixing this is outside the scope of the PR, but I'm just not sure this seems right. Would it make sense to make the if/else at the very top level, always showing connection to users who are not connected - then have the routes under that?

thingalon commented on projects/plugins/boost/app/assets/src/js/elements/Price.svelte:
```diff
@@ -0,0 +1,23 @@
+	const showPriceFraction = ! hidePriceFraction || ! fraction.endsWith( '00' );
+</script>
+
+<p class="jb-price" class:jb-price--not-off-price={! isOff}>
```
This is going to make me unpopular, but I'm looking at the `BenefitsInterstitial` and noting that it's already using the `PricingCard` element from Jetpack. That means we're already loading hundreds of kilobytes of React bloat.

Given that, we may as well use the React component where they're already built.

Given you've got this built, we can roll with this but I think in future we should lean more on premade React components.

thingalon commented on projects/plugins/boost/app/assets/src/js/pages/connection/Connection.svelte:
```diff
@@ -14,10 +13,6 @@
 		__( 'Sell more stuff', 'jetpack-boost' ),
 	];
-
-	$: if ( $connection.connected ) {
-		onConnectionComplete();
-	}
```
I assume this is removed because it's handled in `initialize` in `connection.ts`? Nice. 🧹

thingalon commented on projects/plugins/boost/app/assets/src/js/pages/getting-started/GettingStarted.svelte:
```diff
@@ -0,0 +1,229 @@
+	import externalLinkTemplateVar from '../../utils/external-link-template-var';
+	import { getUpgradeURL } from '../../utils/upgrade';
+
+	const pricing = derived( config, $config => $config.pricing );
```
Something @pyronaur taught me: this can be done like so:
```
$: pricing = $config.pricing;
```

It's easier to read at a glance, and is equally good at keeping `pricing` up to date as `$config` changes, due to `$:`. You can even skip the `let` as Svelte implies that for you. :)

thingalon commented on projects/plugins/boost/app/assets/src/js/pages/getting-started/GettingStarted.svelte:
```diff
@@ -0,0 +1,229 @@
+					<p class="visible-md comparison-table__feature-row">
+						Defer non-essential JavaScript
+						<ContextInfo placement="left">
+							<TemplatedString
+								template={deferJSContext}
+								vars={{
+									...externalLinkTemplateVar( webDevLink ),
+								}}
+							/>
```
I don't think it's worth showing a link in a mouse-over popup, as you can't click on it. :)

haqadn replied:
You can :)

thingalon replied:
Well, count me surprised. You can click on the (i) and it stays up so you can click on the links. Nice!

However, is there a reason the links go to the web.dev homepage, and not to the specific articles via the Jetpack redirects (like on the boost dashboard)?

thingalon commented on projects/plugins/boost/app/assets/src/js/stores/config.ts:
```diff
@@ -22,6 +23,20 @@ export const dismissedPopOuts = {
+export const getStarted = {
+	subscribe: getStartedStore.subscribe,
+	setValue: async ( value: boolean ) => {
+		// eslint-disable-next-line no-console
```
I think this comment is a leftover from removed debug code :)

thingalon commented on projects/plugins/boost/app/assets/src/js/stores/config.ts:
```diff
@@ -22,6 +23,20 @@ export const dismissedPopOuts = {
+const getStartedStore = writable( Jetpack_Boost.site.getStarted );
+export const getStarted = {
+	subscribe: getStartedStore.subscribe,
+	setValue: async ( value: boolean ) => {
```
This method has no reason to be async. I can see it's called from `onConnectionComplete` without an await.

I also think this store has a slightly confusing API. `setValue` forces a flag in local memory which doesn't get pushed to the remote host. `done` sends the updated value to the back-end. I feel like this will lead to confusion in the future.

haqadn replied:
I have removed the separate store and replaced it with usage from directly the config store to make it easier to refresh and use a single source of truth.

thingalon replied:
👍 Nice!

thingalon commented on projects/plugins/boost/app/assets/src/js/utils/connection.ts:
```diff
@@ -11,5 +11,8 @@ export async function onConnectionComplete(): Promise< void > {
 	}
 
+	// Set a flag to indicate that the user has to get started with the plugin.
+	getStarted.setValue( true );
```
🤔 I gather this is setting a local flag to force the user over to the getting started screen? I don't think this will take the wp_options value into account.

Would it be simpler / make more sense to always send the user to the "getting started" screen after connection unless they've already bought Boost premium? That way, we don't have to keep track of a back-end flag, and we have a really simple startup flow.

haqadn replied:
I was going for always requiring the user to select a plan.

If they click Connect, do not wait; move to another page, and come back later, it would still ask the user to choose a plan.

What you suggested makes sense to me too. Let me know which you want to go for.

thingalon replied:
I think the difference between the two approaches is negligible, and will only impact edge-case users. I'm fine to go with your model. :)

thingalon commented on projects/plugins/boost/tests/e2e/lib/pages/wp-admin/JetpackBoostPage.js:
```diff
@@ -25,6 +25,19 @@ export default class JetpackBoostPage extends WpPage {
+	/**
+	 * Select the free plan from getting started page.
+	 */
+	async chooseFreePlan() {
+		const button = await this.page.$( '.comparison-table__cta .jb-button--outline' );
+		await button.click();
+		await this.page.waitForResponse(
+			response =>
+				response.url().match( /jetpack-boost\/v1\/get-started/ ) && response.status() === 200,
+			{ timeout: 2 * 60 * 1000 }
+		);
+	}
```
Nice, thanks for updating the E2E tests along the way :-)

thingalon reviewed (changes requested):
There's a z-order issue with the extra info popups.

![image](https://user-images.githubusercontent.com/1369626/194979677-08953a61-cc4a-409f-b426-45bae9fbe38c.png)

There's also inconsistent tabs/spaces use throughout the scss files. 

Whereabouts is the "Critical CSS" feature turned on by default? I didn't spot it in the code while reviewing it. :)

mikestottuk commented:
Some comments from me

1. The flow currently doesn't enable CSS module by default which was one of the key drivers behind the project (to increase that module infiltration)
2. If I click get started without connecting, I get an error (if I know the URL to go to)
3. On first use, I got `Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'currencyCode')`  and I wasn't connected. Reloading the page fixed it but might be a wider problem.


<img width="1920" alt="Screenshot 2022-10-11 at 10 37 17" src="https://user-images.githubusercontent.com/4077604/195055736-589e3d96-d40a-4759-8bec-a0f4db834058.png">

haqadn committed:
Fixes and improvements

haqadn committed:
Improve tooltip

haqadn committed:
Make spacing on context-info icon smmother

haqadn committed:
Make CSS indentation consistent

haqadn committed:
Improve routing structure and block routes before connection

haqadn committed:
Enable critical CSS while getting started

haqadn commented:
> Some comments from me
> 
> 1. The flow currently doesn't enable CSS module by default which was one of the key drivers behind the project (to increase that module infiltration)
> 2. If I click get started without connecting, I get an error (if I know the URL to go to)
> 3. On first use, I got `Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'currencyCode')`  and I wasn't connected. Reloading the page fixed it but might be a wider problem.
> 
> <img alt="Screenshot 2022-10-11 at 10 37 17" width="1920" src="https://user-images.githubusercontent.com/4077604/195055736-589e3d96-d40a-4759-8bec-a0f4db834058.png">

I have addressed the points you mentioned. @mikemayhem3030

haqadn committed:
Update CCSS generation title

haqadn committed:
Merge branch 'trunk' into add/first-user-flow

thingalon commented on projects/plugins/boost/app/assets/src/js/stores/config.ts:
```diff
@@ -22,6 +23,18 @@ export const dismissedPopOuts = {
 	},
 };
 
+// eslint-disable-next-line camelcase
```
Is this needed? I don't see any non-camel case on the next line :)

haqadn replied:
True, removed

thingalon commented on projects/plugins/boost/app/assets/src/js/stores/config.ts:
```diff
@@ -22,6 +23,18 @@ export const dismissedPopOuts = {
 };
 
+// eslint-disable-next-line camelcase
+export const getStarted = () => {
+	update( store => ( {
+		...store,
+		site: {
+			...store.site,
+			getStarted: false,
+		},
+	} ) );
+	return saveGetStarted();
+};
```
This method is a big improvement on the previous interface. Love it.

However, the name is slightly confusing. You're not getting started then, so calling `getStarted` seems confusing. See this comment in `PurchaseSuccess.svelte`, and how it contrasts the code:
```
		// Disable the getting started page
		getStarted();
```

I'd jokingly suggest `gotStarted`, but that's not clear either. How about `markGetStartedComplete` ?

haqadn replied:
I have added your name as a prefix as requested

oskosk replied:
Addressed in [b64d21e](https://github.com/Automattic/jetpack/pull/26666/commits/b64d21e5b5448752ad575edb2fb7c160b3f683a6)

thingalon commented on projects/plugins/boost/app/assets/src/js/pages/getting-started/GettingStarted.svelte:
```diff
@@ -0,0 +1,229 @@
+	// svelte-ignore unused-export-let - Ignored values supplied by svelte-navigator.
+	export let navigate, location;
+	const chooseFreePlan = async () => {
+		await Promise.all( [ getStarted(), updateModuleState( 'critical-css', true ) ] );
```
🤔 If the user chooses the paid version will it automatically enable Cloud CSS on their way back from the purchase flow?

haqadn replied:
Yes

haqadn committed:
Remove redundant eslint comment

haqadn committed:
Refactor function name for clarity

haqadn committed:
Skip getting started page for user with premium features

thingalon commented:
After signing up using this flow, I see both Local CSS and Cloud CSS options on the settings page. I'm not sure if it is the fault of this PR yet, so I will dig deeper.

![image](https://user-images.githubusercontent.com/1369626/195748154-f44840c0-8d09-4b7c-94bd-64f130f3b00a.png)

thingalon commented:
After purchasing during this flow, there are two critical CSS options, and they both fail:
![image](https://user-images.githubusercontent.com/1369626/195748644-24f52437-7211-4bd3-bee4-e8ab9a8d7f06.png)

haqadn commented:
> After purchasing during this flow, there are two critical CSS options, and they both fail: ![image](https://user-images.githubusercontent.com/1369626/195748644-24f52437-7211-4bd3-bee4-e8ab9a8d7f06.png)

I wasn't able to reproduce this both on local and jurassic ninja. Are you sure you are on the latest build?

thingalon commented:
> I wasn't able to reproduce this both on local and jurassic ninja. Are you sure you are on the latest build?

Ok this one is on me. I had some garbage settings leftover in my local test env from a previous test.

haqadn committed:
Do not enable a performance module if its not available

haqadn committed:
Fix web.dev links

haqadn committed:
Visual improvements on context info bubble

haqadn committed:
Remove discount badge from premium plan

thingalon reviewed (dismissed):
👍 Looks good; nice work 🚢

Commit:
Merge branch 'trunk' into add/first-user-flow

haqadn committed:
[not verified] Fix e2e

haqadn committed:
Fix e2e

haqadn committed:
Enable getting-started page on plugin activation

haqadn committed:
Add tracks on cta click on getting-started page

Commit:
Merge branch 'trunk' into add/first-user-flow

haqadn committed:
[not verified] Style fix

haqadn committed:
[not verified] Improve style of PricingTable react component

haqadn committed:
[not verified] Switch getting started page to use react component

haqadn committed:
[not verified] changelog

Commit:
Merge branch 'trunk' into add/first-user-flow

haqadn commented:
@gmjuhasz, could you please review the changes in the pricing table? I noticed that the `overflow: hidden;` on the `.table` was causing the tooltips to get cut off. So, I have slightly changed the styles to accommodate that. 
![image](https://user-images.githubusercontent.com/3737780/196370701-0fb878b9-60d3-464e-b9fa-dc8219ee98f6.png)

haqadn committed:
Do not use typescript for boost react files to avoid ts errors

haqadn committed:
Fix e2e

haqadn committed:
Fix .is-jb-primary buttons

haqadn committed:
Fix shadows on premium plan on pricing table

haqadn committed:
Fix rating pop-out shadows

haqadn committed:
Remove unused styles and components

Commit:
Remove unnecessary dep

Commit:
Merge branch 'trunk' into add/first-user-flow

Commit:
[not verified] Turn off the tsc rule that imports must be used as values, as the react components we consume don't follow that rule

haqadn committed:
Convert BoostPricingTable back to tsx

thingalon reviewed (approved):
Looks great! Good work @eAdnan007

thingalon commented:
This is good to go IMHO! We just need an approval from @Automattic/jetpack-crew (or someone else with appropriate approve access? @oskosk?) because it touches on a storybook component.

oskosk reviewed (approved):
LGTM!
